src/passes/StringLifting.cpp

```diff
@@ -0,0 +1,218 @@
+        } else if (curr->target == parent.concatImport) {
+          replaceCurrent(
+            Builder(*getModule())
+              .makeStringConcat(curr->operands[0], curr->operands[1]));
+        } else if (curr->target == parent.intoCharCodeArrayImport) {
+          replaceCurrent(Builder(*getModule())
+                           .makeStringEncode(StringEncodeWTF16Array,
+                                             curr->operands[0],
+                                             curr->operands[1],
+                                             curr->operands[2]));
+        } else if (curr->target == parent.equalsImport) {
+          replaceCurrent(Builder(*getModule())
+                           .makeStringEq(StringEqEqual,
+                                         curr->operands[0],
+                                         curr->operands[1]));
+        } else if (curr->target == parent.compareImport) {
+          replaceCurrent(Builder(*getModule())
+                           .makeStringEq(StringEqCompare,
+                                         curr->operands[0],
+                                         curr->operands[1]));
+        } else if (curr->target == parent.lengthImport) {
+          replaceCurrent(
+            Builder(*getModule())
+              .makeStringMeasure(StringMeasureWTF16, curr->operands[0]));
+        } else if (curr->target == parent.charCodeAtImport) {
+          replaceCurrent(
+            Builder(*getModule())
+              .makeStringWTF16Get(curr->operands[0], curr->operands[1]));
+        } else if (curr->target == parent.substringImport) {
+          replaceCurrent(Builder(*getModule())
+                           .makeStringSliceWTF(curr->operands[0],
+                                               curr->operands[1],
+                                               curr->operands[2]));
+        }
+      }
+
+      void visitFunction(Function* curr) {
+        // If we made modifications then we need to refinalize, as we replace
+        // externrefs with stringrefs, a subtype.
+        if (modified) {
+          ReFinalize().walkFunctionInModule(curr, getModule());
+        }
+      }
+    };
+
+    StringApplier applier(*this);
+    applier.run(getPassRunner(), module);
+    applier.walkModuleCode(module);
+
+    // TODO: Add casts. We generate new string.* instructions, and all their
+    //       string inputs should be stringref, not externref, but we have not
+    //       converted all externrefs to stringrefs (since some externrefs might
+    //       be something else). It is not urgent to fix this as the validator
+    //       accepts externrefs there atm, and since toolchains will lower
+    //       strings out at the end anyhow (which would remove such casts). Note
+    //       that if we add a type import for stringref then this problem would
+    //       become a lot simpler (we'd convert that type to stringref).
```

Comment:
Would there be any downside of considering the casts to be part of the semantics of the lifted string operations? We would have to update the interpreter to perform the casts, but this better matches the semantics and typing of the lowered string operations. We would type the output of e.g. string.concat as `string` as an optimization, but it's inputs would remain `extern` references, allowing all valid lowered modules to be represented as lifted modules without having to insert new casts.

Reply:
Interesting, yeah, we could do things that way too.

It does feel "cleaner" though for `string.concat` to receive `stringrefs`... I'd lean toward waiting for type imports and using that to fix this, eventually (as it doesn't seem urgent to fix).

Reply:
We don't need to solve this now because the fuzzer currently ensures that only strings flow into string operations? But that means we're missing out on fuzzing interesting programs that might send other values to string builtins.

Reply:
Hmm, good point... while we import only functions and tags, and we only create strings as true externrefs, we could end up with a struct or array that is externalized. So a stringref could get such a non-string value.

It looks like the fuzzer does not generate new internalize instructions itself. So the reason we have not seen this yet is that hitting this depends on using one of the few initial contents with an internalize, and happening to put exactly the right thing there.

We would need to fix this before adding more internalizes, in other words. Which is worth doing, as you said.